[develop] Add Community Fire Behavior Model

mkavulich opened the pull request:
## DESCRIPTION OF CHANGES: 

This PR introduces the Community Fire Behavior Module (https://github.com/ufs-community/ufs-weather-model/pull/2220) to the SRW App. In addition, there are a number of general improvements to the UFS SRW code and workflow. All of these changes are described in more detail below.

### Community Fire Behavior Module

Documentation has been added to the Users Guide, but the basics are that the fire build requires an additional build flag

```
./devbuild.sh --platform=[hera|derecho|etc.] --app=ATMF
```

and has an additional `fire:` section in `config.yaml`. There is an example fire configuration available in the new example config `config.fire.yaml`. Many settings are described in detail in `config_defaults.yaml`, but I am working on additional details.

The fire model outputs additional netCDF files with fire-specific information; two example plots of this data can be found here:
![ncview emis_smoke_trim](https://github.com/user-attachments/assets/f97ff6b2-764a-40f4-b64b-9a6a68453078)
![ncview fire_area_trim](https://github.com/user-attachments/assets/de126b41-c4ed-4cd8-b392-1e75dcded686)

Sample data for the case in the example file has been staged on Derecho (/glade/derecho/scratch/kavulich/FIRE/2024/test_data) and Hera (/scratch2/BMC/fv3lam/kavulich/UFS/workdir/UFS_FIRE/test_data). I will reach out to EPIC to see if they would be willing to officially stage this sample data on all Level 1 platforms for community use and potential future tests. This PR does not include a WE2E test because the testing system does not currently have a neat way to handle multiple build types, but in theory the new example case could be used. 

Two new pre-defined domains `SUBCONUS_CO_1km` and `SUBCONUS_CO_3km`, for use with high-resolution fire experiments, are added.

### `build_settings.yaml`
In `develop` there is currently no way to reference build settings at configure and run time. There was a vestigial capability in the `ufs_srweather_app.settings.in` which was added 4+ years ago, but it has never really been used or updated and it is in an ad-hoc format that cannot be easily used. I have leveraged this existing capability to create a new file, `build_settings.yaml` in the build directory that contains useful build information (the application built, the compiler, the git hash, etc.) and can be easily read using uwtools in the existing workflows and run scripts. Specifically this is used in this PR to ensure that the app has been built with the fire behavior capability if fire options are selected, but this should prove very useful in other parts of the workflow going forward; for just a few examples:
 - checking that the right compiler was specified for the experiment settings
 - checking that machine settings are correct compared to the build
 - checking app-specific settings for other builds such as AQM

This new file and the unused (but potentially useful in the future) `ufs_srweather_app_meta.h.in` files are moved to the `sorc` directory to keep the source code more organized and uncluttered.

### Improve capability to use a different set of vertical levels
Currently the [instructions for running SRW with a different number of vertical levels from the default 65](https://ufs-srweather-app.readthedocs.io/en/latest/UsersGuide/CustomizingTheWorkflow/LAMGrids.html#changing-the-number-of-vertical-levels) is quite convoluted, involving the need to manually edit the template namelist file. This PR adds LEVP as a configuration variable to more easily allow for runs with different vertical levels. The users guide instructions have been updated accordingly.

### Flexible config file name
Now rather than being forced to use the name `config.yaml` for your experiment config, you can reference any yaml file by using the `-c` option for workflow generation:

```
./generate_FV3LAM_wflow.py -c config.fire.yaml
```

With these changes, `EXPT_CONFIG_FN` is removed as a configurable option, but this was uselessly self-referential anyway.

### Additional options for `retrieve_data.py`
A big limitation in specifying filenames in the current app is that there is no capability for filenames to change based on the forecast hour vs the cycle hour. Therefore a new set of variables, identical to the originals but with the prefix `f`, can now be used when specifying filenames with respect to the forecast hour rather than the cycle hour.

An example of this new usage can be seen in the newly added `config.fire.yaml` experiment:

```
  EXTRN_MDL_FILES_LBCS:
    - '{fyyyymmdd}.hrrr.t{fhh}z.wrfprsf00.grib2'
```
This instructs `retrieve_data.py` to look for filenames specific to the valid forecast hour for the specific lateral boundary condition file. So for forecast hour zero (valid 2020081318), the expected filename will be `20200813.hrrr.t18z.wrfprsf00.grib2`. But at the 18 hour forecast, the LBC filename will be `20200814.hrrr.t12z.wrfprsf00.grib2`. This was not previously possible without these new variables, and represents a vast improvement in flexibility for users.

### Speedup of symbolic linking
I found that (depending on the platform) the `create_symlink_to_file.sh` function could take as many as several seconds for each link, an absurdly long time for such a simple operation! I removed some superfluous logic from this function and `boolify.sh` (which ensures that booleans are properly compared in bash, a necessary evil), which brought this time down to tenths of seconds. This results in a speedup of a minute or more for the run_fcst task, depending on the platform and the settings.

### Various random improvements
 - Fixed regex statements in WE2E tests to fix `SyntaxWarning: invalid escape sequence`
 - Updated build modulefiles to include `FC` variable, which is needed by the Fire Behavior component.

### Type of change
<!-- Please delete options that are not relevant. Add an X to check off a box. -->
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## TESTS CONDUCTED: 
A number of fundamental test runs have been conducted on Derecho, Hera, and Orion, both with and without the ATMF build. Not all are passing, but seems to be unrelated to this development; see the note at the top. This section will be updated once those issues are resolved.

<!-- Explicitly state what tests were run on these changes, or if any are still pending (for README or other text-only changes, just put "None required"). Make note of the compilers used, the platform/machine, and other relevant details as necessary. For more complicated changes, or those resulting in scientific changes, please be explicit! -->
<!-- Add an X to check off a box. -->

- [x] derecho.intel
  - fundamental tests, both with and without -a=ATMF build
  - new UFS_FIRE test
- [ ] gaea.intel
- [ ] hera.gnu
- [ ] hera.intel
- [ ] hercules.intel
- [ ] jet.intel
- [ ] orion.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [ ] fundamental test suite
- [ ] comprehensive tests (specify *which* if a subset was used)

## DEPENDENCIES:
#1136 should be merged first, as it includes a fix for Hera runs with the new UFS weather model hash.

## DOCUMENTATION:
Users guide has been updated with instructions for building and running with these new capabilities; see draft documentation build in the Github checks below. Specifically, [this is the new chapter on the UFS_FIRE capability](https://ufs-srweather-app--1139.org.readthedocs.build/en/1139/UsersGuide/BuildingRunningTesting/FIRE.html).

## ISSUE: 
 - Resolves #1147

## CHECKLIST
<!-- Add an X to check off a box. -->
- [x] My code follows the style guidelines in the Contributor's Guide
- [x] I have performed a self-review of my own code using the Code Reviewer's Guide
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

## CONTRIBUTORS (optional): 
Thanks to the whole Fire Behavior team for bringing me on to this project, and their extensive efforts to get this capability ready for use by the scientific community: @danrosen25 @masih-e @pedro-jm and @mefrediani

danrosen25 committed:
Checkout ufs-weather-model from ESMF for fire comp

mkavulich committed:
First hooks for UFS FIRE added to workflow. Generates experiment successfully, but won't run correctly.

mkavulich committed:
Add capability to specify YAML config file name in generate function

mkavulich committed:
Implement better/more concise jinja templating for nems.configure file,
cleanup create_nems_configure_file.py to use arguments rather than environment variables,
add fire capabilities

mkavulich committed:
Almost there!

 - Add remainder of configurable fire namelist options to config_defaults
 - Add fire namelist filenames and paths to config_defaults
 - Add LEVP (number of vertical levels) as a config variable, and explicitly
   set levp and npz as namelist variables. This will require some updates to documentation
 - Generate the majority of the UFS_FIRE namelist in the workflow generation script. The
   rest (start and end times) will be set in the run_fcst step to account for cycling

mkavulich committed:
Some fixes up through the run_fcst step:
 - Needed correct naming template for input ICS and LBCS in config.fire.yaml
 - Incorrectly formatted lines in create_nems_configure_file.py
 - Fix incorrect namelist location for levp
 - Forgot to ignore FIRE_INPUT_DIR when creating &fire namelist

mkavulich committed:
In the final stretch I think! Edits to run_fcst script:

 - Fill in &time namelist section
 - Link namelist.fire to namelist.input for now (until we can get the name changed)
 - Link in geo_em file

Stuck on namelist errors for now due to old weather model hash.

mkavulich committed:
Fire capability should be False by default

mkavulich committed:
Forgot to commit namelist.fire parm file

mkavulich committed:
Better error (FileNotFoundError) for set_namelist.py if the provided file does not exist. This avoids weird error handling from calling an argparse error class.

mkavulich committed:
- Weather model has been updated, need to remove deprecated options
 - Force bash to output decimal numbers when creating fire namelist
 - Remove rogue ln command

mkavulich committed:
Almost there...

 - Create a new variable FIRE_NUM_TASKS to choose the number of tasks to assign to the fire behavior model.
 - Modify PE_MEMBER01 so it accounts for FIRE_NUM_TASKS if set
 - Check that FIRE_NUM_TASKS is positive if running the fire model
 - Remove last remnants of old namelist variables
 - Update nems.configure to assign weather model to first petlist, then the fire tasks last

mkavulich committed:
Fire run is working! Add some fixes for non-fire runs

mkavulich committed:
QOL improvements for WE2E_summary file

mkavulich committed:
Fix crontab bug

mkavulich committed:
Better defaults for config.fire.yaml

mkavulich committed:
First attempt at two-way coupling workflow

mkavulich committed:
Update nems.configure for new FIRE settings

mkavulich committed:
Update field_table for fire-specific tracer fsmoke

mkavulich committed:
Update weather model branch

mkavulich committed:
Reference config dictionary, not exported variables

mkavulich committed:
Updates for weather model hash 75d28a14:

 - Remove deprecated namelist options "fire_grows_only" and "fire_lfn_ext_up"
 - Add functionality to create_diag_table_file.py that gives the ability to add additional entries to an existing diag_table
 - When UFS_FIRE is active, add line to diag_table for "fsmoke" tracer

mkavulich committed:
Better default values in config.fire.yaml

mkavulich committed:
Merge remote-tracking branch 'upstream/develop' into feature/ufs_fire_workflow

Resolved Conflicts:
	Externals.cfg
	parm/ufs.configure
	ush/create_ufs_configure_file.py
	ush/generate_FV3LAM_wflow.py
	ush/get_crontab_contents.py
	ush/set_namelist.py

mkavulich committed:
Fix merge problems

mkavulich committed:
Fix bug in &time namelist section setup: didn't move temporary file back where it should be!

mkavulich committed:
*ACTUALLY* fix bug in &time namelist section setup

mkavulich committed:
Merge remote-tracking branch 'upstream/develop' into feature/ufs_fire_workflow

mkavulich committed:
Fix merge issues

mkavulich committed:
Cutting the fat out of base utilities boolify and create_symlink_to_file. This gives ~5x speedup for create_symlink_to_file on derecho

mkavulich committed:
More merge corrections

 - No more usage function in create_symlink_to_file()
 - Updates for new uwtools functionality

mkavulich committed:
Fix problem with uw namelist update

mkavulich committed:
Add ability in retrieve_data.py to use forecast hour templates as well (e.g. {fyyyy}, {fhh}, {fyyyymmdd}, etc) to supplement the current templates based on cycle date

mkavulich committed:
Fix fire default filenames

mkavulich committed:
Remove mentions of Cheyenne from comments

mkavulich committed:
Merge remote-tracking branch 'upstream/develop' into feature/ufs_fire_workflow

mkavulich committed:
Update WM to point to main repo and newer hash; this requires an update to build library versions

mkavulich committed:
Need new version of UPP as well to correspond with updated WM hash

mkavulich committed:
Okay, about time we put these build-time settings to use....

 - Rename ufs_srweather_app.settings --> build_settings.yaml, make yaml format for easy reading
 - Add "Application" and "Git hash" to build_settings.yaml
 - Read build_settings.yaml in setup.py to determine if App was built with fire capabilities

mkavulich committed:
Dont use pregenerated ICs in sample case since it needs to work on all platforms

mkavulich committed:
Add FC definition to build modulefiles to fix fire module build issue

mkavulich committed:
Fix syntax warning for bad regular expression in WE2E test script

mkavulich committed:
Fix problems with non-fire cases; opened issue about non-functional threading in current develop

mkavulich committed:
Fixes for OMP threading problem from Michael

mkavulich committed:
Merge remote-tracking branch 'upstream/develop' into feature/ufs_fire_workflow

mkavulich committed:
Move new build_settings_template to sorc/ directory

mkavulich committed:
Move ufs_srweather_app_meta.h.in to sorc/

mkavulich committed:
Update instructions for custom vertical levels; I completely forgot I made this improvement months ago!

mkavulich committed:
Update fire case for shorter forecast, and to use new variable syntax

mkavulich committed:
Additional consistency checks in setup.py, remove set -x from run_fcst

mkavulich committed:
Make the linter happy

mkavulich committed:
Fix long line in create diag table script, set build settings to only print in debug mode

mkavulich committed:
Fix linting for generate_FV3LAM_wflow.py. Really wish there was a way to run these tests locally...

mkavulich committed:
Manage externals takes forever and rarely fails, so run that test last

mkavulich committed:
Update diag_table test

mkavulich committed:
We don't want to hide output from CI unit tests in case there's a failure

mkavulich committed:
Need dummy build settings file for test_generate_FV3LAM_wflow.py now

mkavulich committed:
Fix the linter for the test. This is an extremely annoying process!

mkavulich committed:
Forgot you need externals for the unit tests

mkavulich committed:
Add new fire WE2E test; will only work on Derecho for now. Also move "aqm" directory to "special_builds" directory to go with fire

mkavulich committed:
Fix error introduced when trying to please linter

mkavulich committed:
Add Fire documentation to Users Guide

mkavulich committed:
Details about output; having trouble embedding image though

mkavulich committed:
Add broken image for demonstration

christinaholtNOAA commented on devbuild.sh:
```diff
@@ -16,7 +16,7 @@ OPTIONS
   -a, --app=APPLICATION
       weather model application to build; for example, ATMAQ for Online-CMAQ
-      (e.g. ATM | ATMAQ | ATMW | S2S | S2SW)
+      (e.g. ATM | ATMAQ | ATMF | ATMW | S2S | S2SW)
```
Should this list be pruned to the ones the app supports? 

Do you want to add a blurb about ATMF in the text above the list?

mkavulich replied:
Done

christinaholtNOAA commented on doc/UsersGuide/BuildingRunningTesting/FIRE.rst:
```diff
@@ -0,0 +1,250 @@
+
+The `Community Fire Behavior Model (CFBM) <https://ral.ucar.edu/model/community-fire-behavior-model>`_ is a wildland fire model coupled to the UFS Atmospheric Model. The capability to run this code is now available in the UFS Short-Range Weather App for easy use by the community. The `fire_behavior repository <https://github.com/NCAR/fire_behavior>`_ is a :term:`submodule` of the UFS Weather Mode ...
+
+The biggest difference between the UFS FIRE capability and other modes of the UFS SRW is that a special build flag is required to build the coupled fire behavior code, as described in the instructions below. Aside from that, additional input files, and some fire-specific config settings, building and running is the same as any other use of SRW.
```
```suggestion
The biggest difference between the UFS FIRE capability and other modes of the UFS SRW is that a special build flag is required to build the coupled fire behavior code, as described in the instructions below. Aside from that, the need for additional input files, and some fire-specific config settings, configuring and running an experiment is the same as any other use of SRW.
```

christinaholtNOAA commented on doc/UsersGuide/BuildingRunningTesting/FIRE.rst:
```diff
@@ -0,0 +1,250 @@
+
+   ./manage_externals/checkout_externals
+
+Build the SRW App with Fire Behavior enabled
```
```suggestion
Build the SRW App with Fire Behavior Enabled
```

christinaholtNOAA commented on doc/UsersGuide/BuildingRunningTesting/FIRE.rst:
```diff
@@ -0,0 +1,250 @@
+Load the |wflow_env| Environment
+--------------------------------------------
+
+Load the python environment for the workflow:
```
```suggestion
Load the appropriate modules for the workflow:
```

This step also loads the Rocoto module.

christinaholtNOAA commented on doc/UsersGuide/BuildingRunningTesting/FIRE.rst:
```diff
@@ -0,0 +1,250 @@
+.. code-block:: console
+
+   cd ush
+   cp config.fire.yaml config.yaml 
```
This is no longer necessary, right?

mkavulich replied:
I'd like to keep this as-is; it's a good idea to not work on top of the example file since that is versioned in the repository.

christinaholtNOAA commented on doc/UsersGuide/BuildingRunningTesting/FIRE.rst:
```diff
@@ -0,0 +1,250 @@
+
+.. code-block:: console
+
+   ./generate_FV3LAM_wflow.py
```
This gets a command line argument now, no?

mkavulich replied:
It's not mandatory so I don't want to clutter up this specific doc too much, but I will update the docs elsewhere for this new option.

christinaholtNOAA commented on doc/UsersGuide/BuildingRunningTesting/FIRE.rst:
```diff
@@ -0,0 +1,250 @@
+   fire_output_2020-08-13_18:30:00.nc  fire_output_2020-08-13_20:00:00.nc  fire_output_2020-08-13_21:30:00.nc  fire_output_2020-08-13_23:00:00.nc
+   fire_output_2020-08-13_19:00:00.nc  fire_output_2020-08-13_20:30:00.nc  fire_output_2020-08-13_22:00:00.nc  fire_output_2020-08-13_23:30:00.nc
+
+   These files contain output directly from the fire model (hence why they are at a greater frequency), including variables such as the fire perimeter and area, smoke emitted, and fuel percentage burnt. 
```
```suggestion
   These files contain output directly from the fire model (hence the greater frequency), including variables such as the fire perimeter and area, smoke emitted, and fuel percentage burned. 
```

christinaholtNOAA reviewed (changes requested):
As promised, you have added a lot of documentation. I only got as far as FIRE.rst in this first round, but will return to it early next week.

These are only minor suggestions. Do with them what you will.

mkavulich committed:
Merge remote-tracking branch 'upstream/develop' into feature/ufs_fire_workflow

mkavulich committed:
Fix default number of cores; fire cores should not be multiplied by OMP

mkavulich committed:
Don't specify locations of data for WE2E test

mkavulich committed:
Fix merge for ufs.configure file

mkavulich commented:
@MichaelLueken I will need help staging the data for the new UFS_FIRE WE2E test. It is located on Hera at `/scratch2/BMC/fv3lam/kavulich/UFS/workdir/UFS_FIRE/test_data/geo_em.d01.nc` and `/scratch2/BMC/fv3lam/kavulich/UFS/workdir/UFS_FIRE/test_data/HRRR/`; feel free to organize the data in a way that makes sense in the EPIC test data space and I can change the test files to point to it accordingly.

mkavulich committed:
Minor fixes to FIRE chapter: Remove irrelevant sentence about tasks, fix indentation, fix intra-chapter link

mkavulich committed:
Fixed image of fire output

MichaelLueken commented on doc/UsersGuide/CustomizingTheWorkflow/ConfigWorkflow.rst:
Thanks for removing the `EXPT_CONFIG_FN` parameter from the documentation!  Please add the new `LEVP` and `UFS FIRE` configuration parameters to the `ConfigWorkflow.rst` file as well.

MichaelLueken reviewed (changes requested):
@mkavulich -

I was able to successfully run the new UFS Fire sample configuration on Derecho.

It would be best to include the new `UFS Fire` configuration parameters, as well as `LEVP`, in the updates to `doc/UsersGuide/CustomizingTheWorkflow/ConfigWorkflow.rst`.

christinaholtNOAA commented on parm/namelist.fire:
```diff
@@ -0,0 +1,40 @@
+ end_day = 14
+ end_hour = 21
+ end_minute = 00
+ end_second = 00
```
UW Tools will likely be able to help you remove these hard-coded values with references to a Python datetime object. Rendering the template will allow you to avoid the octal problem you mention in the bash script below.

Here, the changes might look like:

```
&time
  start_year = '{{ cycle.year }}'
  start_month = '{{ cycle.month }}'
  start_day = '{{ cycle.day }}'
  ...
  end_year = '{{ endtime.year }}'
/
```


To render this to stdout, you can use:
```
uw template render --input-file fire.nml --values-format yaml --values-file <(echo '{cycle: !!timestamp "2024-09-10 12:00:00", endtime: !!timestamp "2024-09-15 12:00:00"}')
```

mkavulich replied:
Per our offline discussion, I wasn't able to get this working due to a potential `uw template` issue. I did remove the time values from the base namelist however.

christinaholtNOAA commented on parm/ufs.configure:
```diff
@@ -3,63 +3,47 @@
-  ATM
+{% for line in runseq -%}
+{{ line }}
+{%- endfor %}
```
I'm assuming this bit replaces the run sequence stuff deleted above? Nice.

mkavulich replied:
That's correct; the run sequence for each component is now specified in `create_ufs_configure_file.py`

christinaholtNOAA commented on scripts/exregional_run_fcst.sh:
```diff
@@ -756,6 +755,49 @@ fi
+  /
+"
+
+  echo $settings | uw config realize --update-format nml --input-format nml --output-format nml --input-file "${FIRE_NML_FP}" -o "${FIRE_NML_FN}"
```
Here is where you might change to `uw template render` suggested above, but use the `-o "${FIRE_NML_FN}"` to have it go to the appropriate file instead of stdout.

This would require a change to the error handling below, too.

christinaholtNOAA commented on tests/WE2E/WE2E_summary.py:
```diff
@@ -6,7 +6,13 @@
+    print("\n\nERROR: Could not load python utilities.")
+    print('Note that this script can only be run in the SRW App from the directory:')
+    print("ufs-srweather-app/tests/WE2E\n\n")
+    raise
```
Could this not be simplified by using a UW config tool from the API?

mkavulich replied:
Most likely, but incorporating uw tools into the WE2E tests is beyond the scope of this PR

christinaholtNOAA commented on tests/WE2E/WE2E_summary.py:
```diff
@@ -68,9 +74,13 @@ def setup_logging(debug: bool = False) -> None:
-    # Calculate core hours and update yaml
-    expts_dict = calculate_core_hours(expts_dict)
-    write_monitor_file(yaml_file,expts_dict)
+    if expts_dict:
```
I know this was already here with this name, but somehow putting it in an if-statement like this, I now read "expects dict". Not sure if it matters much.

christinaholtNOAA commented on tests/WE2E/test_configs/special_builds/config.UFS_FIRE.yaml:
```diff
@@ -0,0 +1,60 @@
+metadata:
```
I'm not sure you want to carve out the space "special builds" for this. We'd presumably have many different "ufs-weather-model app builds" in the future. Maybe it makes sense to put it under a subdir `test_configs/ufs_fire/`, then we may have several test configs for the UFS Fire build, and others under similarly named `test_configs/mpas_dycore`, or `test_configs/aqm`, etc.

mkavulich replied:
Your reasoning makes sense, I'll revert this change.

christinaholtNOAA commented on tests/WE2E/test_configs/special_builds/config.UFS_FIRE.yaml:
```diff
@@ -0,0 +1,60 @@
+fire:
+  UFS_FIRE: True
+  FIRE_NUM_TASKS: 1
+  FIRE_INPUT_DIR: /glade/derecho/scratch/kavulich/FIRE/2024/test_data
```
I assume this would eventually live under the EPIC role account space on the Tier I platforms?

mkavulich replied:
Yes, I believe EPIC is working on staging the data in the appropriate locations; I will update the test file as necessary when that is complete.

christinaholtNOAA commented on ush/config_defaults.yaml:
```diff
@@ -1918,7 +1928,7 @@ task_run_fcst:
   PRINT_ESMF: false
 
-  PE_MEMBER01: '{{ OMP_NUM_THREADS_RUN_FCST * (LAYOUT_Y * LAYOUT_X + WRTCMP_write_groups * WRTCMP_write_tasks_per_group) if QUILTING else OMP_NUM_THREADS_RUN_FCST * (LAYOUT_Y * LAYOUT_X)}}'
+  PE_MEMBER01: '{{ fire.FIRE_NUM_TASKS + OMP_NUM_THREADS_RUN_FCST * (LAYOUT_Y * LAYOUT_X + WRTCMP_write_groups * WRTCMP_write_tasks_per_group) if QUILTING else OMP_NUM_THREADS_RUN_FCST * (LAYOUT_Y * LAYOUT_X) + fire.FIRE_NUM_TASKS }}'
```
Just to keep it clearer, would you mind changing it so that the `fire.FIRE_NUM_TASKS` entry ends up in the same position in both of these equations? I personally prefer seeing it show up at the end, but consistency is preferable either way.

christinaholtNOAA commented on ush/create_ufs_configure_file.py:
```diff
@@ -62,35 +96,42 @@ def create_ufs_configure_file(run_dir):
     #
-    # Call a python script to generate the experiment's actual UFS_CONFIG_FN
-    # file from the template file.
+    # Call uwtools "render" set_template function from workflow_tools to fill in jinja template
+    # from config template and settings variable to create ufs.configure file for this experiment
```
```suggestion
    # Call the uwtools "render" function to fill in jinja expressions contained in the UFS Configure file template
    # with the values from the settings variable to create ufs.configure file for this experiment
```

Would you mind using the term `jinja expressions` here? It would make it more consistent with uwtools documentation and may help users who are cross-referencing docs. In general, we've been trying to refer to the `{{ }}` entries as Jinja expressions, and then the file that contains them would be a template file.

christinaholtNOAA commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -770,6 +831,8 @@ def setup_logging(logfile: str = "log.generate_FV3LAM_wflow", debug: bool = Fals
                      "according to the parameters specified in the config file\n")
 
+    parser.add_argument('-c', '--config', default='config.yaml',
+                        help='Name of experiment config file in YAML format')
```
🎉 🥂 👍 🥇

christinaholtNOAA commented on ush/predef_grid_params.yaml:
```diff
@@ -334,9 +401,6 @@
 #
-# Note:
-# This grid has not been thoroughly tested (as of 20201027).
-#
```
I'm assuming this shows up in the test suite now?

mkavulich replied:
Yes, the test is `grid_RRFS_AK_3km_ics_FV3GFS_lbcs_FV3GFS_suite_HRRR`

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -1488,6 +1506,35 @@ def _dict_find(user_dict, substring):
+    fire_conf = expt_config["fire"]
+    if fire_conf["UFS_FIRE"]:
+        if build_config["Application"]!="ATMF":
+            raise Exception("UFS_FIRE == True but UFS SRW has not been built for fire coupling; see users guide for details")
```
Nifty check!

christinaholtNOAA reviewed (commented):
This is a really nice addition of a new feature and some great upgrades for SRW.

mkavulich committed:
Apply suggestions from code review

Co-authored-by: Christina Holt <56881914+christinaholtNOAA@users.noreply.github.com>

mkavulich committed:
Make the linter happy;

mkavulich committed:
Better description for -a build flag

mkavulich committed:
Revert move of aqm test directory, make new "fire" test directory, rename fire test to indicate its one-way coupled

mkavulich committed:
Consistent order in PE_MEMBER01 definition

mkavulich committed:
Instead of a separate entry for all these duplicated namelist settings, make all the ignition-specific variables lists of length FIRE_NUM_IGNITIONS

mkavulich committed:
Update build_settings to include "Machine" value, add check for machine consistency in setup.py, convert some logging calls to specific logger object

mkavulich committed:
- Update fire test for new setting names; add new multi-fire test
 - Fix missing values for fire namelist logic in workflow generation

mkavulich committed:
Need to convert fire namelist settings to a list if not provided as a list

mkavulich committed:
Update fire instructions in users guide for latest commits; add descriptions of config options

mkavulich committed:
Initial commit of fire capability does not support two-way coupling

mkavulich committed:
Add LEVP setting documentation

mkavulich commented:
@MichaelLueken I have added documentation of the config.yaml options for the new Fire capabilities and LEVP. In the future it might be nice to consolidate this information, as it is currently duplicated across the Users guide and `config_defaults.yaml`.

mkavulich committed:
Remove default time values from namelist.fire

mkavulich committed:
Fix generate workflow unit test

mkavulich commented:
@christinaholtNOAA @MichaelLueken I believe I have addressed all your comments. Note that I did update how the fire ignition variables are handled in `setup.py` and `generate_FV3LAM_wflow.py` slightly, and add a new WE2E test that covers multiple ignitions. You can see those changes in [this diff](https://github.com/ufs-community/ufs-srweather-app/pull/1139/files/85ca43eb8ee2939198c3281ec30bec774c73dafa..d4043bd2e36748d37ee9883e990f3271d2241f61).

Let me know if you have any more comment/concerns.

mkavulich committed:
Fix test description and running instructions

MichaelLueken commented on ush/setup.py:
```diff
@@ -1484,14 +1509,66 @@ def _dict_find(user_dict, substring):
+                    logger.critical(f"{fire_conf[setting]=}")
+                    raise ValueError(f"For FIRE_NUM_IGNITIONS > 1, {setting} must be a list of the same length")
+
+        if fire_conf["FIRE_ATM_FEEDBACK"] > 0.0:
+            raise ValueError("FIRE_ATM_FEEDBACK > 0 (two-way coupling) not supported in UFS yet")
```
The new WE2E tests and the `config.fire.yaml` sample configuration set `FIRE_ATM_FEEDBACK` to `1.0`.  This is causing the tests to fail since it is larger than `0.0`.  Should the test configurations be updated to use `FIRE_ATM_FEEDBACK: 0.0`, rather than `1.0`?

mkavulich replied:
Sorry about that, I updated the settings in the WE2E tests but failed to push it here. This is an issue that we found after merging the capability into the weather model where the two-way coupling was not actually working (results were identical to 1-way coupling); this just ensures that users don't try to use two-way coupling while it's not working in the initial version.

MichaelLueken replied:
That makes sense to me!  This has been resolved as of [e67173f](https://github.com/ufs-community/ufs-srweather-app/pull/1139/commits/e67173fb206c3d38bf78318b8eebfa5c4d0be035).

MichaelLueken reviewed (commented):
@mkavulich - While attempting to run the two new WE2E tests, they failed to submit because `FIRE_ATM_FEEDBACK > 0.0` for these new configurations.  Should the test configuration use `FIRE_ATM_FEEDBACK: 0.0` instead of `FIRE_ATM_FEEDBACK: 1.0`?

mkavulich committed:
Update config.fire.yaml

Initial capability needs FIRE_ATM_FEEDBACK=0.0

mkavulich committed:
Update config.UFS_FIRE_multifire_one-way-coupled.yaml

FIRE_ATM_FEEDBACK = 0.0

mkavulich committed:
Update config.UFS_FIRE_one-way-coupled.yaml

FIRE_ATM_FEEDBACK = 0.0

MichaelLueken reviewed (approved):
Thanks, @mkavulich!  The tests are successfully generated and run.  Also, thank you very much for updating the documentation in `ConfigWorkflow.rst`.  We'll need to see if there is some way to pull information directly from `config_defaults.yaml` into the `ConfigWorkflow.rst` file so that we don't need to repeat modifications to both.

mkavulich commented:
@MichaelLueken Thanks for the review. Let me know when the data is staged for the WE2E tests; I'll likely need to make some modifications to those test files to point to the correct data.

MichaelLueken commented:
@mkavulich -

The UFS Fire data has been staged on Tier-1 platforms:
- Hera - /scratch1/NCEPDEV/nems/role.epic/UFS_SRW_data/develop/UFS_FIRE
- Derecho - /glade/work/epicufsrt/contrib/UFS_SRW_data/develop/UFS_FIRE
- Gaea - /gpfs/f5/epic/world-shared/UFS_SRW_data/develop/UFS_FIRE
- Hercules/Orion - /work/noaa/epic/role-epic/contrib/UFS_SRW_data/develop/UFS_FIRE
- Jet - /mnt/lfs5/HFIP/hfv3gfs/role.epic/UFS_SRW_data/develop/UFS_FIRE

And many thanks to @RatkoVasic-NOAA for staging the data across the platforms!

mkavulich committed:
Point to official CFBM users guide location now that its published

mkavulich committed:
Final updates for fire tests

 - New variable to WE2E test data, we should use this to consolidate test variables in the future
 - Shorten fire tests to 3 hours
 - Fix extra prints in workflow generation

mkavulich commented:
@MichaelLueken I pushed the final changes for the new fire WE2E tests, I believe this PR is ready for merging presuming all tests pass and @christinaholtNOAA approves.

christinaholtNOAA reviewed (approved):
LGTM.

MichaelLueken commented:
@mkavulich -

The Jenkins tests are all failing in the `Functional WorkflowTaskTests` phase of the pipeline.  The Jenkins tests don't have an `exec` directory, but an `install_${compiler}/exec` directory.  The `ush/setup.py` script needs to have the `build_config_fp` variable updated to point to either `exec` (standard build) or `install_${compiler}/exec` (Jenkins build), otherwise the Jenkins tests will be unable to run.

MichaelLueken commented:
@mkavulich -

On Hera, you can see the directory structure for GNU - `/scratch2/NAGAPE/epic/role.epic/jenkins/workspace/s-srweather-app_pipeline_PR-1139/hera` - and Intel - `/scratch2/NAGAPE/epic/role.epic/jenkins/workspace/s-srweather-app_pipeline_PR-1139__2/hera`.  We'll need to ensure that the new `build_settings.yaml` file can be located, whether the file is in `exec` or in `some_user_defined_directory/exec`.

mkavulich committed:
Fix for jenkins test exec directory location

mkavulich commented:
@MichaelLueken Thanks for the update. I pushed a fix that should work, can you try again?

MichaelLueken commented:
Thanks, @mkavulich!  The Jenkins tests have been resubmitted.

MichaelLueken commented:
@mkavulich -

The Jenkins tests are now running without issue, but there is a conflict in `doc/conf.py` where Gillian removed the `Options for autoyaml extension` section that you had modified in this PR.

MichaelLueken commented:
@mkavulich -

All Jenkins WE2E tests have successfully passed!  I will be able to merge this PR once the conflict in `doc/conf.py` has been resolved.

mkavulich committed:
Merge branch 'develop' into feature/ufs_fire_workflow

mkavulich committed:
Fix link to fire UG
